File: py/2024/day01b.py
def parseInput() -> tuple[dict[int, int], list[int]]:
    similarity: dict[int, int] = {}
    left_numbers: list[int] = []
    right_numbers: list[int] = []

    for line in open(0):
        left, right, *rest = line.split()

        if left not in similarity:
            similarity[int(left)] = 0

        left_numbers.append(int(left))
        right_numbers.append(int(right))

    for number in right_numbers:
        if number in left_numbers:
            similarity[number] += 1

    return similarity, left_numbers


if __name__ == "__main__":
    total: int = 0

    similarity, left_numbers = parseInput()
    for number in left_numbers:
        total += number * similarity[number]

    print(total)

File: go/2024/day01b.go
//go:build ignore

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	fmt.Println(solve())
}

func solve() int {
	similarity := make(map[int]int)
	var left_numbers []int

	scanner := bufio.NewScanner(bufio.NewReader(os.Stdin))
	for scanner.Scan() {
		numbers := strings.Fields(scanner.Text())

		left, err := strconv.Atoi(numbers[0])
		if err != nil {
			fmt.Println(err)
		}

		left_numbers = append(left_numbers, left)

		right, err := strconv.Atoi(numbers[1])
		if err != nil {
			fmt.Println(err)
		}

		if _, ok := similarity[right]; !ok {
			similarity[right] = 1
		} else {
			similarity[right]++
		}
	}

	total := 0
	for _, v := range left_numbers {
		if _, ok := similarity[v]; ok {
			total += v * similarity[v]
		}
	}

	return total
}
